### PR TITLE
Upgrade to LLVM 14.0.2

### PR DESCRIPTION
**Pull request by vchuravy**
TODO:
- [x] analyzegc
- [x] llvmpasses
- [x] Win64 Build
- [x] ASAN
- [x] 8X performance `foldl`
- [x] Mac/Codegen failure
```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-x64-5.0/build/default-macmini-x64-5-0/julialang/julia-master/julia-538dba5996/share/julia/test/compiler/codegen.jl:272
  Expression: was_gced
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -192,7 +192,7 @@ else
 JL_PRIVATE_LIBS-$(USE_SYSTEM_ZLIB) += libz
 endif
 ifeq ($(USE_LLVM_SHLIB),1)
-JL_PRIVATE_LIBS-$(USE_SYSTEM_LLVM) += libLLVM libLLVM-13jl
```
Comment (pchintalapudi) [inline Makefile:195]: I believe a similar changes is needed in src/Makefile

**Comment by pchintalapudi** [comment]
I think the current build error can be fixed by moving `#define DEBUG_TYPE` lower down in our optimization files (probably DomTreeBuilder defines and undefs its own DEBUG_TYPE).

**Comment by vchuravy** [comment]
Win64 looks like a fun one:

```
 cd /cygdrive/c/buildbot/worker/package_win64/build/base && /cygdrive/c/buildbot/worker/package_win64/build/usr/bin/julia.exe -C "generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)" --output-ji `cygpath -w /cygdrive/c/buildbot/worker/package_win64/build/usr/lib/julia/corecompiler.ji`.tmp --startup-file=no --warn-overwrite=yes -g0 -O0 compiler/compiler.jl
Assertion failed: inserted, file /cygdrive/c/buildbot/worker/package_win64/build/src/jitlayers.cpp, line 763
```

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45195#issuecomment-1123823494) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/aebcebc_vs_13ae079/report.md).

**Comment by pchintalapudi** [comment]
foldl seems to be the biggest regression here (~8X), and some of the summation benchmarks seem to have improved for floats.

**Comment by gbaraldi** [comment]
That big of change sounds like a failure to SIMD

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45195#issuecomment-1123825075) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b3febcb_vs_13ae079/report.html).

**Comment by vchuravy** [comment]
The aarch64 failure is known:

```
      From worker 4:	julia: /workspace/srcdir/llvm-project/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp:507: void llvm::RuntimeDyldELF::resolveAArch64Relocation(const llvm::SectionEntry&, uint64_t, uint64_t, uint32_t, int64_t): Assertion `isInt<33>(Result) && "overflow check failed for relocation"' failed.
      From worker 4:	
      From worker 4:	signal (6): Aborted
```

It's the reason we switched to JITLink for aarch64-darwin. cc: @lhames @weliveindetail I will happily offer support if the GSoC student wants use Julia as a stress-test for JITLink on Elf

**Comment by vchuravy** [comment]
```
compiler/codegen                          (3) |         failed at 2022-05-21T21:12:00.891
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-x64-5.0/build/default-macmini-x64-5-0/julialang/julia-master/julia-538dba5996/share/julia/test/compiler/codegen.jl:272
  Expression: was_gced
 
```
on aarch64-darwin and

```
signal (11): Segmentation fault
in expression starting at none:0
jl_is_submodule at /workspace/srcdir/src/module.c:901
module_in_worklist at /workspace/srcdir/src/dump.c:182 [inlined]
jl_collect_backedges at /workspace/srcdir/src/dump.c:1233 [inlined]
ijl_save_incremental at /workspace/srcdir/src/dump.c:2682
jl_write_compiler_output at /workspace/srcdir/src/precompile.c:65
ijl_atexit_hook at /workspace/srcdir/src/init.c:204
jl_repl_entrypoint at /workspace/srcdir/src/jlapi.c:711
main at /workspace/srcdir/cli/loader_exe.c:59
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x400808)
Allocations: 11073626 (Pool: 11067760; Big: 5866); GC: 15
ERROR: LoadError: Failed to precompile Colors [5ae59095-9a9b-59fe-a467-6f913c188581] to /home/pkgeval/.julia/compiled/v1.9/Colors/jl_nAutxB.
```
from PkgEval seem to be the biggest issues. Both are GC related.

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by giordano** [comment]
I couldn't reproduce the failure in `compiler/codegen` the other day on aarch64-apple-darwin, and it also disappeared in https://github.com/JuliaLang/julia/commit/39a24eb0d0cce7f47219775dc0d975e3b4f2ebf7, there is only a failure in backtrace: https://buildkite.com/julialang/julia-master/builds/12262#859b5f1a-17df-4654-887d-e9a571cc21f7
```
The global RNG seed was 0x8b87a2e86632cb974477241986867844.
 
Error in testset backtrace:
Test Failed at /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/julia-master/julia-39a24eb0d0/share/julia/test/backtrace.jl:182
  Expression: found
Error in testset backtrace:
Test Failed at /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/julia-master/julia-39a24eb0d0/share/julia/test/backtrace.jl:194
  Expression: found
ERROR: LoadError: Test run finished with errors
in expression starting at /Users/administrator/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-1.0/build/default-macmini-aarch64-1-0/julialang/julia-master/julia-39a24eb0d0/share/julia/test/runtests.jl:93
ERROR: A test has failed. Please submit a bug report (https://github.com/JuliaLang/julia/issues)
including error messages above and the output of versioninfo():
Julia Version 1.9.0-DEV.634
Commit 39a24eb0d0c (2022-05-22 22:04 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.3 (ORCJIT, apple-m1)
  Threads: 4 on 4 virtual cores
```

**Review by vtjnash — APPROVED** [review]

**Comment by vchuravy** [comment]
@vtjnash I am stuck on the Windows test somehow a linux path creeps into the backtrace.

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45195#issuecomment-1153033897) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/7000b46_vs_a9e3cc7/report.md).

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45195#issuecomment-1153033864) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1c5858b_vs_a9e3cc7/report.html).

**Comment by KristofferC** [comment]
Maybe want to run PkgEval with assertions for PRs like this?

**Comment by vchuravy** [comment]
> Maybe want to run PkgEval with assertions for PRs like this?

That would be a great option to have!

**Comment by vchuravy** [comment]
@DilumAluthge @tkf and I chatted about ASAN and we decided that for now we should make it "allowed to fail" best to keep the "can we build" bit though.

**Comment by DilumAluthge** [comment]
> > Maybe want to run PkgEval with assertions for PRs like this?
> 
> That would be a great option to have!

I believe we already have this option. Control-F https://github.com/JuliaCI/Nanosoldier.jl/blob/master/README.md for "assert".

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = "%self", buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"])`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45195#issuecomment-1153367877) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1c5858b_vs_1c5858b/report.html).

**Comment by KristofferC** [comment]
Feels like this should give some comment about all the assertion failures in the PkgEval report.

**Comment by vchuravy** [comment]
> Feels like this should give some comment about all the assertion failures in the PkgEval report.

... I am blind. We really need to improve the formatting (and order of the report). I looked at the report on my phone and mentally lumped the process aborted and segmentation fault reports under "Networking related" since the heading got completely lost. Sorry about that.

**Comment by gbaraldi** [comment]
We should probably run PkgEval with assertions periodically to catch some assertion specific errors earlier.

**Pull request merged**